### PR TITLE
Handle negative indexes for arrays in fortran

### LINKED ISSUE #248: Conversion of negative indexes in arrays
Currently, `a[-1]` is translated literally to `a(-1)` in Fortran, causing an out-of-bounds warning, and probably not intended behavior. Would it make sense to change it to `a(ubound(a,dim))`? See https://stackoverflow.com/questions/20986269/how-do-i-access-last-item-in-an-array-in-fortran/20986963

### PR DESCRIPTION
**Pull request by EmilyBourne**
Handle negative constants as indexes for arrays. Also handle other indexes that could be negative if the new `allow_negative_index` decorator is specified. This fixes #248 

- Handle indexes which are negative constants by subtracting them from the shape
- Add `allow_negative_index` decorator and save property to `Variable`
- Add modulo operator when the index is not an integer and the user has added the `allow_negative_index` decorator for this variable
- Add tests

### CODE REVIEW

File: pyccel/parser/syntactic.py [hunk 1]
```diff
@@ -632,6 +632,13 @@ def _visit_FunctionDef(self, stmt):
             for i, arg in enumerate(args):
                 args[i] = str(arg).replace("'", '')
             decorators['stack_array'] = tuple(args)
+
+        if 'allow_negative_index' in decorators:
+            args = list(decorators['allow_negative_index'].args)
+            for i, arg in enumerate(args):
+                args[i] = str(arg).replace("'", '')
```
Comment (nhamidn) [inline pyccel/parser/syntactic.py:639]: i didn't understand why using `.replace` here

Comment (EmilyBourne) [reply pyccel/parser/syntactic.py]: You are right. I copied and pasted from the `stack_array` decorator. I suspect it is a hangover from the old parser

**Review by nhamidn — APPROVED** [review]

File: pyccel/parser/syntactic.py [hunk 2]
```diff
@@ -630,8 +630,15 @@ def _visit_FunctionDef(self, stmt):
         if 'stack_array' in decorators:
             args = list(decorators['stack_array'].args)
             for i, arg in enumerate(args):
-                args[i] = str(arg).replace("'", '')
+                args[i] = str(arg)
             decorators['stack_array'] = tuple(args)
```
Comment (yguclu) [inline pyccel/parser/syntactic.py:634]: If there is no need to perform any substitution in the arguments' names, we should rewrite this as
```python
decorators['stack_array'] = tuple(str(a) for a in decorators['stack_array'].args)
```

File: pyccel/parser/syntactic.py [hunk 3]
```diff
@@ -630,8 +630,15 @@ def _visit_FunctionDef(self, stmt):
...
+
+        if 'allow_negative_index' in decorators:
+            args = list(decorators['allow_negative_index'].args)
+            for i, arg in enumerate(args):
+                args[i] = str(arg)
+            decorators['allow_negative_index'] = tuple(args)
```
Comment (yguclu) [inline pyccel/parser/syntactic.py:640]: I have the same comment as above. We should rewrite this as
```python
decorators['allow_negative_index'] = tuple(str(a) for a in decorators['allow_negative_index'].args)
```

**Review by yguclu — CHANGES_REQUESTED** [review]

**Review by yguclu — APPROVED** [review]

**Comment by yguclu** [comment]
This should be the next PR to go through. So please abort the test builds and update the branch from master.

**Comment by yguclu** [comment]
Never mind, I've just done it myself :wink:

**Pull request merged**